pkgs/stdenv/adapters.nix

```diff
@@ -201,7 +201,10 @@ rec {
       inherit bintools;
     };
     allowedRequisites =
-      lib.mapNullable (rs: rs ++ [ bintools pkgs.mold (lib.getLib pkgs.mimalloc) (lib.getLib pkgs.openssl) ]) (stdenv.allowedRequisites or null);
+      (lib.mapNullable (rs: rs) (stdenv.allowedRequisites or null))
```

Comment:
@Artturin  I have an evaluation error in my personal project when using the following snippet of code:

```
ccacheStdenv = pkgs.ccacheStdenv.override {
    stdenv = pkgs.stdenvAdapters.useMoldLinker llvmStdenv;
};
```

<details>
<summary>Trace</summary>

```
~> nix build -L .#devShells.x86_64-linux.default --no-eval-cache --show-trace
error:
       … while calling a functor (an attribute set with a '__functor' attribute)

         at /nix/store/jgnwm94vqil0kn88fv32db6kicimp19z-source/flake.nix:31:22:

           30|       # };
           31|       ccacheStdenv = pkgs.ccacheStdenv.override {
             |                      ^
           32|         stdenv = pkgs.useMoldLinker llvmStdenv;

       … from call site

         at /nix/store/jj02kapbd0nhwjrxpqc2mypa8vpw62v4-source/lib/trivial.nix:430:7:

          429|     { # TODO: Should we add call-time "type" checking like built in?
          430|       __functor = self: f;
             |       ^
          431|       __functionArgs = args;

       … while calling anonymous lambda

         at /nix/store/jj02kapbd0nhwjrxpqc2mypa8vpw62v4-source/lib/customisation.nix:82:32:

           81|       # Re-call the function but with different arguments
           82|       overrideArgs = copyArgs (newArgs: makeOverridable f (overrideWith newArgs));
             |                                ^
           83|       # Change the result of the function call by applying g to it

       … from call site

         at /nix/store/jj02kapbd0nhwjrxpqc2mypa8vpw62v4-source/lib/customisation.nix:82:41:

           81|       # Re-call the function but with different arguments
           82|       overrideArgs = copyArgs (newArgs: makeOverridable f (overrideWith newArgs));
             |                                         ^
           83|       # Change the result of the function call by applying g to it

       … while calling 'makeOverridable'

         at /nix/store/jj02kapbd0nhwjrxpqc2mypa8vpw62v4-source/lib/customisation.nix:72:24:

           71|   */
           72|   makeOverridable = f: origArgs:
             |                        ^
           73|     let

       … from call site

         at /nix/store/jj02kapbd0nhwjrxpqc2mypa8vpw62v4-source/lib/customisation.nix:74:16:

           73|     let
           74|       result = f origArgs;
             |                ^
           75|

       … while calling anonymous lambda

         at /nix/store/jj02kapbd0nhwjrxpqc2mypa8vpw62v4-source/pkgs/top-level/all-packages.nix:18997:44:

        18996|
        18997|   ccacheStdenv = lowPrio (makeOverridable ({ stdenv, ... } @ extraArgs:
             |                                            ^
        18998|     overrideCC stdenv (buildPackages.ccacheWrapper.override ({

       … from call site

         at /nix/store/jj02kapbd0nhwjrxpqc2mypa8vpw62v4-source/pkgs/top-level/all-packages.nix:18998:5:

        18997|   ccacheStdenv = lowPrio (makeOverridable ({ stdenv, ... } @ extraArgs:
        18998|     overrideCC stdenv (buildPackages.ccacheWrapper.override ({
             |     ^
        18999|       inherit (stdenv) cc;

       … while calling 'overrideCC'

         at /nix/store/jj02kapbd0nhwjrxpqc2mypa8vpw62v4-source/pkgs/stdenv/adapters.nix:35:24:

           34|   # Override the compiler in stdenv for specific packages.
           35|   overrideCC = stdenv: cc: stdenv.override { allowedRequisites = null; cc = cc; };
             |                        ^
           36|

       … from call site

         at /nix/store/jgnwm94vqil0kn88fv32db6kicimp19z-source/flake.nix:32:18:

           31|       ccacheStdenv = pkgs.ccacheStdenv.override {
           32|         stdenv = pkgs.useMoldLinker llvmStdenv;
             |                  ^
           33|       };

       … while calling 'useMoldLinker'

         at /nix/store/jj02kapbd0nhwjrxpqc2mypa8vpw62v4-source/pkgs/stdenv/adapters.nix:192:19:

          191|
          192|   useMoldLinker = stdenv: let
             |                   ^
          193|     bintools = stdenv.cc.bintools.override {

       … from call site

         at /nix/store/jj02kapbd0nhwjrxpqc2mypa8vpw62v4-source/lib/trivial.nix:430:7:

          429|     { # TODO: Should we add call-time "type" checking like built in?
          430|       __functor = self: f;
             |       ^
          431|       __functionArgs = args;

       … while calling anonymous lambda

         at /nix/store/jj02kapbd0nhwjrxpqc2mypa8vpw62v4-source/lib/customisation.nix:82:32:

           81|       # Re-call the function but with different arguments
           82|       overrideArgs = copyArgs (newArgs: makeOverridable f (overrideWith newArgs));
             |                                ^
           83|       # Change the result of the function call by applying g to it

       … from call site

         at /nix/store/jj02kapbd0nhwjrxpqc2mypa8vpw62v4-source/lib/customisation.nix:82:41:

           81|       # Re-call the function but with different arguments
           82|       overrideArgs = copyArgs (newArgs: makeOverridable f (overrideWith newArgs));
             |                                         ^
           83|       # Change the result of the function call by applying g to it

       … while calling 'makeOverridable'

         at /nix/store/jj02kapbd0nhwjrxpqc2mypa8vpw62v4-source/lib/customisation.nix:72:24:

           71|   */
           72|   makeOverridable = f: origArgs:
             |                        ^
           73|     let

       … from call site

         at /nix/store/jj02kapbd0nhwjrxpqc2mypa8vpw62v4-source/lib/customisation.nix:74:16:

           73|     let
           74|       result = f origArgs;
             |                ^
           75|

       … while calling anonymous lambda

         at /nix/store/jj02kapbd0nhwjrxpqc2mypa8vpw62v4-source/pkgs/stdenv/generic/default.nix:3:1:

            2|
            3| argsStdenv@{ name ? "stdenv", preHook ? "", initialPath
             | ^
            4|

       … from call site

         at /nix/store/jj02kapbd0nhwjrxpqc2mypa8vpw62v4-source/pkgs/stdenv/generic/default.nix:83:5:

           82|     derivation (
           83|     lib.optionalAttrs (allowedRequisites != null) {
             |     ^
           84|       allowedRequisites = allowedRequisites

       … while calling 'optionalAttrs'

         at /nix/store/jj02kapbd0nhwjrxpqc2mypa8vpw62v4-source/lib/attrsets.nix:679:5:

          678|     # The attribute set to return if `cond` is `true`.
          679|     as:
             |     ^
          680|     if cond then as else {};

       error: value is null while a list was expected
```
</details>

I've investigated this, and it seems that `lib.mapNullable` gets called on a `null` value, hence resulting in a `null`, and then the `++` operator fails because of that. I would suggest changing this line to `stdenv.allowedRequisites or [ ]`.